server/data-service/src/main/java/edu/illinois/ncsa/incore/service/data/controllers/DatasetController.java

```diff
@@ -1048,17 +1048,67 @@ public Object updateObject(@Parameter(name = "Dataset Id from data service", req
+        String propVal = JsonUtils.extractValueFromJsonString(UPDATE_OBJECT_VALUE, inDatasetJson);
+
+        // Handle special logic for sourceDataset update (i.e., setting parent ID)
+        if ("sourceDataset".equalsIgnoreCase(propName)) {
+            Dataset parentDataset = repository.getDatasetById(propVal);
+            if (parentDataset == null) {
+                throw new IncoreHTTPException(Response.Status.BAD_REQUEST, "Parent dataset not found: " + propVal);
+            }
+
+            // Check if the parent dataset is compatible with the child dataset for join
+            String childFormat = dataset.getFormat();
+            String parentFormat = parentDataset.getFormat();
+
+            boolean isChildCompatible = true;
+            boolean isParentCompatible = true;
+            if (childFormat == null || !childFormat.equalsIgnoreCase("table")) {
+                isChildCompatible = false;
+            }
+            if ( parentDataset.getFormat() == null || !parentDataset.getFormat().equalsIgnoreCase("shapefile")) {
+                isParentCompatible = false;
+            }
+
+            // check compatibility of the datasets by GUID
+            boolean isGuidCompatible = ServiceUtils.validateJoinCompatibilityByGuid(dataset, parentDataset, repository);
+
+            dataset.setSourceDataset(propVal);
+            Dataset updatedDataset = repository.updateDataset(datasetId, "sourceDataset", propVal);
+
+            // GeoServer logic (only if format is 'table')
+            if ( isChildCompatible && isParentCompatible && isGuidCompatible) {
+                String format = updatedDataset.getFormat();
+                if (format != null && format.equalsIgnoreCase("table")) {
+                    try {
+                        File geoPkgFile = FileUtils.joinShpTable(updatedDataset, repository, true);
+                        boolean success = GeoserverUtils.uploadGpkgToGeoserver(updatedDataset.getId(), geoPkgFile);
+                        if (!success) {
+                            throw new IncoreHTTPException(Response.Status.INTERNAL_SERVER_ERROR, "GeoServer upload failed.");
+                        }
+                        FileUtils.deleteTmpDir(geoPkgFile); // Clean up
+                        logger.info("Dataset joined to source dataset and uploaded to geoserver.");
+                    } catch (Exception e) {
+                        logger.error("Error during GeoServer upload process: " + e.getMessage(), e);
+                        throw new IncoreHTTPException(Response.Status.INTERNAL_SERVER_ERROR, "GeoServer layer creation failed.");
+                    }
+                }
+            } else {
+                logger.info("Dataset not joined to parent dataset.");
+            }
+            return updatedDataset;
+        }
 
         try {
             Field f = dataset.getClass().getDeclaredField(propName); // Get the passed field from Dataset class
```

Comment:
It would be better if the fieldcheck on line 1114 happened first to make sure the field exists. Then after this check, we can do the extra check for "sourceDataset" as the propName, something like:

if "sourceDataset".equalsIgnoreCase(propName)
     // your code to handle the update to the geoserver
else {
   dataset = repository.updateDataset(datasetId, propName, propVal);
}
return dataset;